addons/point_of_sale/controllers/main.py

```diff
@@ -84,3 +85,125 @@ def print_sale_details(self, date_start=False, date_stop=False, **kw):
         pdf, _ = request.env['ir.actions.report'].with_context(date_start=date_start, date_stop=date_stop)._render_qweb_pdf('point_of_sale.sale_details_report', r)
         pdfhttpheaders = [('Content-Type', 'application/pdf'), ('Content-Length', len(pdf))]
         return request.make_response(pdf, headers=pdfhttpheaders)
+
+    @http.route(['/pos/ticket/validate'], type='http', auth="public", website=True, sitemap=False)
+    def show_ticket_validation_screen(self, access_token='', **kwargs):
+        def _parse_additional_values(fields, kwargs):
+            res = {}
+            for field in fields:
+                if field.name in kwargs:
+                    val = kwargs.pop(field.name)
+                    res[field.name] = val
+            return res
+
+        # If the route is called directly, return a 404
+        if not access_token:
+            return request.not_found()
+        # Get the order using the pos reference and access token. We can't use the id in the route because we may not have it yet when the QR code is generated.
+        pos_order = request.env['pos.order'].sudo().search([
+            ('access_token', '=', access_token),
+        ])
+        if not pos_order:
+            return request.not_found()
+
+        # If the order was already invoiced, return the invoice directly by forcing the access token so that the non-connected user can see it.
+        if pos_order.account_move and pos_order.account_move.is_sale_document():
+            return request.redirect('/my/invoices/%s?access_token=%s' % (pos_order.account_move.id, pos_order.account_move._portal_ensure_token()))
+
+        # Get the optional extra fields that could be required for a localisation.
+        additional_partner_fields = request.env['res.partner'].get_partner_localisation_fields_required_to_invoice(pos_order.company_id.account_fiscal_country_id)
+        additional_invoice_fields = request.env['account.move'].get_invoice_localisation_fields_required_to_invoice(pos_order.company_id.account_fiscal_country_id)
+
+        # Validate the form by ensuring required fields are filled and the VAT is correct.
+        form_values = {'error': {}, 'error_message': {}, 'extra_field_values': {}}
+        if kwargs and request.httprequest.method == 'POST':
+            form_values.update(kwargs)
+            # Extract the additional fields values from the kwargs now as they can't be there when validation the regular form.
+            values = _parse_additional_values(additional_partner_fields, kwargs)
+            form_values['extra_field_values'].update(values)
+            # Do the same for invoice values, separately as they are only needed for the invoice creation
+            invoice_values = _parse_additional_values(additional_invoice_fields, kwargs)
+            form_values['extra_field_values'].update(invoice_values)
```

Comment:
if you're mixing partner and invoice fields in the kwargs, what happens when a l10n module decides to have the same field name on both objects? there's no rule saying it can't happen. You could namespace the parameters with a prefix to differentiate them, or group them in separate container parameters.